### PR TITLE
Refactor docs / Maintenance of docs

### LINKED ISSUE #1430: Split up README.md + Update Docs Config
README.md is getting very large. Let’s split up the components documentation to start with:
- blackd
- black-primer
- editor plugins

Please link from README.md to the specific documentation in the repo or the readthedocs equivalent.

Totally open to other ideas here too.

### LINKED ISSUE #1435: [Question] [Awaiting PR] How do you build the documentation?
So, while working on #1430, I decided learning how to build the documentation was probably a good idea so I can see whether I break the docs. The problem is that I have never used Sphinx. So after doing some quick Google searches, I ran `make html` in the docs directory. Which errored out:

```console
PS R:\programming\black\docs> ./make html
Running Sphinx v2.4.4
making output directory... done
c:\users\user\.virtualenvs\black-rhkux7ap\lib\site-packages\sphinx\util\compat.py:30: RemovedInSphinx30Warning: The config variable "source_parsers" is deprecated. Please update your extension for the parser and remove the setting.
  warnings.warn('The config variable "source_parsers" is deprecated. '
c:\users\user\.virtualenvs\black-rhkux7ap\lib\site-packages\sphinx\util\compat.py:36: RemovedInSphinx30Warning: app.add_source_parser() does not support suffix argument. Use app.add_source_suffix() instead.
  app.add_source_parser(suffix, parser)
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 24 source files that are out of date
updating environment: [new config] 24 added, 0 changed, 0 removed
c:\users\user\.virtualenvs\black-rhkux7ap\lib\site-packages\recommonmark\parser.py:75: UserWarning: Container node skipped: type=document
  warn("Container node skipped: type={0}".format(mdnode.t))
reading sources... [ 37%] flake_bugbear
Exception occurred:
  File "c:\users\user\.virtualenvs\black-rhkux7ap\lib\site-packages\recommonmark\parser.py", line 225, in visit_list
    self.current_node.append(list_node)
AttributeError: 'NoneType' object has no attribute 'append'
The full traceback has been saved in C:\Users\User\AppData\Local\Temp\sphinx-err-ne49li9r.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```
Then, I updated Sphinx to the latest version and tried again. While the docs did build successfully, it wasn't usable:
image
I also experienced 366 warnings during the build, which seems *excessive*. Most of them were `toctree contains reference to document 'authors' that doesn't have a title: no link will be generated` warnings, and (almost) the rest was Markdown warnings.

[log.txt](https://github.com/psf/black/files/4653366/log.txt)
(warning: it's more than 1000 lines long)

I also tried installing `Sphinx==1.7.2` and `recommonmark==0.4.0` listed in `docs/requirements.txt`, which didn't work as well.

In short, **what on earth am I doing wrong here** and **how you do build the _Black_ documentation**?

**Environment:**
* OS version: `Windows 10 Home Edition Build 18363`
* Python version: `Python 3.8.1`
* _Black_ version: `19.10b1.dev100+gb0f3798`

---

Sorry about the unnecessary noise. I just figured that someone in the _Black_ community would know how to build the documentation. This is probably a stupid question. Feel free to close this issue if needed.

### PR DESCRIPTION
**Pull request by ichard26**
Resolves #1430.
Related #990 (please see #1457).
Resolves #1435.

**See the latest version of the refactored docs here:**
https://ichard26-testblackdocs.readthedocs.io/en/refactor_docs/index.html
(also is proof that this won't break the ReadTheDocs builds)

### Refactor

- Split these sections from the README into their own files as per #1430:
  - `blackd` -> `docs/blackd.md`
  - `black-primer` -> `docs/black_primer.md`
  - `Editor integration` -> `docs/editor_integration.md`
  - `The Black code style` as per @zsol https://github.com/psf/black/issues/1430#issuecomment-630208621 -> `docs/the_black_code_style.md`
  - `Pragmatism` as per @zsol https://github.com/psf/black/issues/1430#issuecomment-631729659 -> `docs/the_black_code_style.md`

~~Yes, `Pragmatism` and `The Black code style` were merged into `docs/the_black_code_style_and_pragmatism.md`. **Please suggest a better title and filename for this poor piece of documentation.**~~ 

Thank you @hugovk for the suggestions! `the_black_code_style_and_pragmatism.md` has been renamed to `the_black_code_style.md`.

### Maintenance

- Fixed a few broken or likely to be broken links:
  - `README.md`:
    - `https://poetry.eustace.io/` -> `https://python-poetry.org/`
    - `https://docs.pipenv.org/` -> `https://readthedocs.org/projects/pipenv/`
  - `docs/editor_integration.md`
    - `https://github.com/proofit404/blacken` -> `https://github.com/pythonic-emacs/blacken`
- Deleted all of the symlinks in `docs/` since they serve no use now:
  - *you don't need me to list them... right?*
- Manually merged these changes since the docs that they were changing have been moved:
  - https://github.com/psf/black/commit/bb767863b0cc51897f9cd94da84b8b01f3936607
  - https://github.com/psf/black/commit/a2408b3cb23fe252b3674cee484d742496bb3411
  - https://github.com/psf/black/commit/397b16431e7e2af2869353d9310834f8c80e6cde
- Committed my changes from #1436.

### Features / Improvements

- ~~Made the sidebar navigation scrollable since we have so much packed into that sidebar that some of its contents are clipped. The only sad part of this change is that for users of Firefox, they will have to see an ugly scrollbar since Firefox disabled hiding scrollbar by default~~ (please see #1457)
- Completely reworked the logic in `docs/conf.py`.
  - Added a `DocSection` dataclass which represents sections of documentation
    - The `name` attribute is the section's name.
    - The `src` attribute is where to get the section's contents.
    - The `processors` attribute is processors to call on the section's contents before writing them back. This extensibility feature can be used to implement a link autoreplacer for example :)
    - The `out_filename` attribute is the filename to be used when writing the section. Only the merged `the_black_code_style_and_pragmatism` section uses this feature.
    - The `src_range` attribute is well... a `src_range` object
    - The `get_out_filename` method returns the `out_filename` attribute if set. If not, then it returns the section's name + `.md`
  - Added a `SrcRange` dataclass which represents a line range of the section's contents
  - Added a `get_contents` function which takes a `DocSection` and returns the contents of that section
  - Moved the header fixing code from the now deleted `generate_sections_from_readme` function into a processor which is ran for all sections generated from `README.md`.
  - Completely redone the `generate_sections_from_readme` function and renamed it to `get_sections_from_readme`. It now spits out a list of `DocSection` objects.
  - Created a new function called `process_sections` which takes `custom_sections` and `readme_sections` and processes them. The workflow of how it works is listed below:
    - Extend `custom_sections` with `readme_sections`
    - While iterating through each `DocSection` from the now extended `custom_sections`:
      - Check if a section which has the same name has been processed already. If so, then skip processing that section. This is how `custom_sections` overrides the default sections from `readme_sections`.
      - Get the target_path via `target_path: Path = CURRENT_DIR / section.get_out_filename()`
      - Get the section's contents using `get_contents(section)`
      - Run all processors defined from `section.processors` over the section's contents
      - Write the section's contents to `target_path`
      - Append the section's name to `processed_sections` so duplicates can be detected and skipped
- Using the new logic, I defined a `custom_sections` list which contains `DocSection` objects. This defines the custom / non-readme-based sections:
<details>

<summary>custom_sections</summary>

```py3
custom_sections = [
    DocSection("the_black_code_style", CURRENT_DIR / "the_black_code_style.md",),
    DocSection("pragmatism", CURRENT_DIR / "the_black_code_style.md",),
    DocSection("editor_integration", CURRENT_DIR / "editor_integration.md"),
    DocSection("blackd", CURRENT_DIR / "blackd.md"),
    DocSection("black_primer", CURRENT_DIR / "black_primer.md"),
    DocSection("contributing_to_black", CURRENT_DIR / ".." / "CONTRIBUTING.md"),
    DocSection("change_log", CURRENT_DIR / ".." / "CHANGES.md"),
]
```

</details>

### CODE REVIEW

File: docs/black_primer.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+- Using `-k` / `--keep` + `-w` / `--work-dir` so you don't have to re-checkout the repo
+  each run
+
+## CLI Arguments
+
+```text
+Usage: black-primer [OPTIONS]
+
+  primer - prime projects for blackening ... 🏴
```
Comment (hugovk) [inline docs/black_primer.md:34]: From https://github.com/psf/black/commit/a2408b3cb23fe252b3674cee484d742496bb3411:
```suggestion
  primer - prime projects for blackening... 🏴
```

File: docs/black_primer.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  -c, --config PATH      JSON config file path  [default: /Users/cooper/repos/
+                         black/src/black_primer/primer.json]
+
+  --debug                Turn on debug logging  [default: False]
+  -k, --keep             Keep workdir + repos post run  [default: False]
+  -L, --long-checkouts   Pull big projects to test  [default: False]
+  -R, --rebase           Rebase project if already checked out  [default:
+                         False]
+
+  -w, --workdir PATH     Directory Path for repo checkouts  [default: /var/fol
```
Comment (hugovk) [inline docs/black_primer.md:46]:
```suggestion
  -w, --workdir PATH     Directory path for repo checkouts  [default: /var/fol
```

File: docs/black_primer.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+explains each parameter:
+
+```json
+{
+  "projects": {
+    "00_Example": {
+      "cli_arguments": "List of extra CLI arguments to pass Black for this project",
+      "expect_formatting_changes": "Boolean to indicate that the version of Black is expected to cause changes",
+      "git_clone_url": "URL you would pass `git clone` to check out this repo",
+      "long_checkout": "Boolean to have repo skipped by defauult unless `--long-checkouts` is specified",
```
Comment (hugovk) [inline docs/black_primer.md:66]:
```suggestion
      "long_checkout": "Boolean to have repo skipped by default unless `--long-checkouts` is specified",
```

File: docs/black_primer.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  -k, --keep             Keep workdir + repos post run  [default: False]
+  -L, --long-checkouts   Pull big projects to test  [default: False]
+  -R, --rebase           Rebase project if already checked out  [default:
+                         False]
+
+  -w, --workdir PATH     Directory Path for repo checkouts  [default: /var/fol
+                         ders/tc/hbwxh76j1hn6gqjd2n2sjn4j9k1glp/T/primer.20200
+                         517125229]
+
+  -W, --workers INTEGER  Number of parallel worker coroutines  [default: 69]
```
Comment (hugovk) [inline docs/black_primer.md:50]:
```suggestion
  -W, --workers INTEGER  Number of parallel worker coroutines  [default: 2]
```

File: docs/conf.py [hunk 5]
```diff
@@ -14,28 +14,70 @@
...
+    end_line: int
+
+
+@dataclass
+class DocSection:
+    """Tracks information about a section of documentation.
+
+    Data:
+        name: The section's name. This will used to detect duplicate sections.
+        src: The filepath to get its contents
```
Comment (hugovk) [inline docs/conf.py:50]:
```suggestion
        src: The filepath to get its contents.
```

**Comment by hugovk** [comment]
> Yes, `Pragmatism` and `The Black code style` were merged into `docs/the_black_code_style_and_pragmatism.md`. **Please suggest a better title and filename for this poor piece of documentation.**

Perhaps:

* The _Black_ code style
* `docs/style.md`

File: docs/black_primer.md (hunk 1 above)
Comment (ichard26) [reply docs/black_primer.md:34]: I thought I manually merged that in, but it appears I didn't. Good catch!

**Comment by ichard26** [comment]
Sorry that this is taking so long. I did a git merge master into refactor_docs and now I regret not doing  a rebase. This interactive rebase is not fun, full of merge conflicts!

**Comment by ichard26** [comment]
Alright, I've finished what I've wanted to do and resolved the many merge conflicts along the way. The key changes I made are the following:
* Make the ordering and default values of the attributes of the `DocSection` dataclass make a bit more sense (hopefully)
* Added logging so you know what the `process_sections` logic is doing
* Renamed `the_black_code_style_and_pragmatism.md` to `the_black_code_style`
  * and... renamed the nested `The Black code style` sub-header to just `Code style`
* Added @hugovk suggestions, thank you!

File: README.md [hunk 6]
```diff
@@ -1427,6 +538,7 @@ Multiple contributions by:
 - [Pablo Galindo](mailto:Pablogsal@gmail.com)
 - [Peter Bengtsson](mailto:mail@peterbe.com)
 - pmacosta
+- Richard Si
```
Comment (cooperlees) [inline README.md:541]: 🤩 - Docs superstar

**Pull request merged**